libbeat/api/server.go

```diff
@@ -19,86 +19,77 @@ package api
+	config Config
 }
 
-func rootHandler() func(http.ResponseWriter, *http.Request) {
-	return func(w http.ResponseWriter, r *http.Request) {
-		// Return error page
-		if r.URL.Path != "/" {
-			http.NotFound(w, r)
-			return
-		}
-
-		w.Header().Set("Content-Type", "application/json; charset=utf-8")
-
-		data := monitoring.CollectStructSnapshot(monitoring.GetNamespace("info").GetRegistry(), monitoring.Full, false)
-
-		print(w, data, r.URL)
+// New creates a new API Server.
+func New(log *logp.Logger, mux *http.ServeMux, config *common.Config) (*Server, error) {
+	if log == nil {
+		log = logp.NewLogger("")
 	}
-}
 
-// stateHandler reports state metrics
-func stateHandler(w http.ResponseWriter, r *http.Request) {
-	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	cfg := DefaultConfig
+	err := config.Unpack(&cfg)
+	if err != nil {
+		return nil, err
+	}
 
-	data := monitoring.CollectStructSnapshot(monitoring.GetNamespace("state").GetRegistry(), monitoring.Full, false)
+	l, err := makeListener(cfg)
+	if err != nil {
+		return nil, err
+	}
 
-	print(w, data, r.URL)
+	return &Server{mux: mux, l: l, config: cfg, log: log.Named("api")}, nil
 }
 
-// statsHandler report expvar and all libbeat/monitoring metrics
-func statsHandler(w http.ResponseWriter, r *http.Request) {
-	w.Header().Set("Content-Type", "application/json; charset=utf-8")
-
-	data := monitoring.CollectStructSnapshot(monitoring.GetNamespace("stats").GetRegistry(), monitoring.Full, false)
-
-	print(w, data, r.URL)
+// Start starts the HTTP server and accepting new connection.
+func (s *Server) Start() {
+	s.log.Info("Starting stats endpoint")
+	go func(l net.Listener) {
+		s.log.Infof("Metrics endpoint listening on: %s (configured: %s)", l.Addr().String(), s.config.Host)
+		http.Serve(l, s.mux)
+		s.log.Infof("Finished starting stats endpoint: %s", l.Addr().String())
+	}(s.l)
 }
 
-func datasetHandler(w http.ResponseWriter, r *http.Request) {
-	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+// Stop stops the API server and free any resource associated with the process like unix sockets.
+func (s *Server) Stop() error {
+	return s.l.Close()
+}
 
-	data := monitoring.CollectStructSnapshot(monitoring.GetNamespace("dataset").GetRegistry(), monitoring.Full, false)
+func parse(host string, port int) (string, string, error) {
+	url, err := url.Parse(host)
+	if err != nil {
+		return "", "", err
+	}
 
-	print(w, data, r.URL)
-}
+	// When you don't explicitely define the Scheme we fallback on tcp + host.
+	if len(url.Host) == 0 && len(url.Scheme) == 0 {
+		addr := host + ":" + strconv.Itoa(port)
+		return "tcp", addr, nil
+	}
 
-func print(w http.ResponseWriter, data common.MapStr, u *url.URL) {
-	query := u.Query()
-	if _, ok := query["pretty"]; ok {
-		fmt.Fprintf(w, data.StringToPrint())
-	} else {
-		fmt.Fprintf(w, data.String())
+	switch url.Scheme {
+	case "http":
+		return "tcp", url.Host, nil
+	case "unix":
+		return url.Scheme, url.Path, nil
```

Comment:
Not sure if we should care about this but just want to check.

Using `-E http.host=unix:///tmp/mb.sock` and `curl -v --unix-socket /tmp/mb.sock 'http://unix:5434/'` works just fine. Should we return an error if the scheme is `unix` and a port is set as well?